### ISSUE TITLE
Encoding issue

### ISSUE THREAD

**Issue by cilerler** [ExcelDataReader/ExcelDataReader#51]
One of my file has an encoding issue. First screenshot shows the original Excel view.  2nd one is via LinqPad, 3rd one is via Visual Studio watch tab.  Any idea how to solve this issue?  Thx

```
const string worksheet = "BD- 1";
const string fileName = "199811021.xls";
const int rowNumber = 3;
const int columnNumber = 8;

string directory = Environment.GetFolderPath(Environment.SpecialFolder.Desktop);
string filePath = Path.Combine(directory, fileName);

FileStream stream = File.Open(filePath, FileMode.Open, FileAccess.Read);
IExcelDataReader excelReader = ExcelReaderFactory.CreateBinaryReader(stream);
DataSet result = excelReader.AsDataSet();

foreach (DataTable t in result.Tables)
{
    if (t.TableName.Equals(worksheet))
    {
        DataRow r = t.Rows[rowNumber];
        object c = r[columnNumber]; 

        // outputs
        r.Dump();
        Console.WriteLine (c);
    }
}

excelReader.Close();
```

![1-excel](https://cloud.githubusercontent.com/assets/385958/5701411/cd97b6e6-9a18-11e4-80b8-d4ec971ad76d.png)
![2-linqpad](https://cloud.githubusercontent.com/assets/385958/5701482/ead99ee4-9a19-11e4-9e23-a9cd56273e7f.png)
![3-visualstudio](https://cloud.githubusercontent.com/assets/385958/5701412/cd97a714-9a18-11e4-81b6-cd0b642933e7.png)

**Comment by wizardgsz** [comment]
Maybe you've to handle Encoding like ExcelBinaryReader does. Perhaps a new propery to the Interface could help...

**Comment by cilerler** [comment]
Thanks but I tried to change it in source (to overwrite/force that part) before I post it here.  I believe I'm doing something wrong so it could be great to implement

**Comment by cilerler** [comment]
Thanks @wizardgsz but I believe it already has the encoding property in it.  And as far as I can see, it detects correct code-page and encoding itself but not applying it. To prove that, I just extended it with following logic which basically allows me to force encoding.  Then I tried to see the results for each encoding from the GetEncodings()  Unfortunately it always returns the same result which is odd.  Any idea what I'm missing here?

You may download the [XLS file here](http://ilerler.com/-excelReader/199811021.xls)

**in public static class ExcelReaderFactory**

        public static IExcelDataReader CreateBinaryReader(Stream fileStream, Encoding encoding)
        {
            IExcelDataReader reader = new ExcelBinaryReader(encoding);
            reader.Initialize(fileStream);

            return reader;
        }

**in public class ExcelBinaryReader**

	private readonly bool forceEncoding

	internal ExcelBinaryReader(Encoding encoding) : this()
	{
		m_encoding = encoding;
		forceEncoding = true;
	}

	private void readWorkBookGlobals()
	{
	//  ...
		case BIFFRECORDTYPE.CODEPAGE:
			m_globals.CodePage = (XlsBiffSimpleValueRecord)rec;                        
			try
			{
				if (!forceEncoding)
				m_encoding = Encoding.GetEncoding(m_globals.CodePage.Value);
			}
	//  ...
	}

**and running the code as**

		void Main()
		{
			foreach (EncodingInfo ei in Encoding.GetEncodings())
			{
				Run(ei.GetEncoding());
			}
		}

		public static void Run(Encoding encoding)
        {
            const string worksheet = "BD- 1";
			const string fileName = "199811021.xls";
			const int rowNumber = 3;
			const int columnNumber = 8;
			
			string directory = Environment.GetFolderPath(Environment.SpecialFolder.Desktop);
			string filePath = Path.Combine(directory, fileName);
			
			FileStream stream = File.Open(filePath, FileMode.Open, FileAccess.Read);
			IExcelDataReader excelReader = ExcelReaderFactory.CreateBinaryReader(stream, encoding);
			DataSet result = excelReader.AsDataSet();
			
			foreach (DataTable t in result.Tables)
			{
				if (t.TableName.Equals(worksheet))
				{
					DataRow r = t.Rows[rowNumber];
					object c = r[columnNumber]; 
			
					// outputs
					r.Dump();
					Console.WriteLine (c);
				}
			}
			
			excelReader.Close();
        }

**Comment by ExcelDataReader** [comment]
You might try getting the develop branch version. There were a few changes
related to encoding in that.

On Tue, Jan 27, 2015 at 11:47 AM, Cengiz Ilerler <notifications@github.com>
wrote:

> Thanks @wizardgsz <https://github.com/wizardgsz> but I believe it already
> has the encoding property in it. And as far as I can see, it detects
> correct code-page and encoding itself but not applying it. To prove that, I
> just extended it with following logic which basically allows me to force
> encoding. Then I tried to see the results for each encoding from the
> GetEncodings() Unfortunately it always returns the same result which is
> odd. Any idea what I'm missing here?
>
> You may download the XLS file here
> <http://ilerler.com/-excelReader/199811021.xls>
>
> *in public static class ExcelReaderFactory*
>
>     public static IExcelDataReader CreateBinaryReader(Stream fileStream, Encoding encoding)
>     {
>         IExcelDataReader reader = new ExcelBinaryReader(encoding);
>         reader.Initialize(fileStream);
>
>         return reader;
>     }
>
> *in public class ExcelBinaryReader*
>
> private readonly bool forceEncoding
>
> internal ExcelBinaryReader(Encoding encoding) : this()
> {
>     m_encoding = encoding;
>     forceEncoding = true;
> }
>
> private void readWorkBookGlobals()
> {
> //  ...
>     case BIFFRECORDTYPE.CODEPAGE:
>         m_globals.CodePage = (XlsBiffSimpleValueRecord)rec;
>         try
>         {
>             if (!forceEncoding)
>             m_encoding = Encoding.GetEncoding(m_globals.CodePage.Value);
>         }
> //  ...
> }
>
> *and running the code as*
>
>     void Main()
>     {
>         foreach (EncodingInfo ei in Encoding.GetEncodings())
>         {
>             Run(ei.GetEncoding());
>         }
>     }
>
>     public static void Run(Encoding encoding)
>     {
>         const string worksheet = "BD- 1";
>         const string fileName = "199811021.xls";
>         const int rowNumber = 3;
>         const int columnNumber = 8;
>
>         string directory = Environment.GetFolderPath(Environment.SpecialFolder.Desktop);
>         string filePath = Path.Combine(directory, fileName);
>
>         FileStream stream = File.Open(filePath, FileMode.Open, FileAccess.Read);
>         IExcelDataReader excelReader = ExcelReaderFactory.CreateBinaryReader(stream, encoding);
>         DataSet result = excelReader.AsDataSet();
>
>         foreach (DataTable t in result.Tables)
>         {
>             if (t.TableName.Equals(worksheet))
>             {
>                 DataRow r = t.Rows[rowNumber];
>                 object c = r[columnNumber];
>
>                 // outputs
>                 r.Dump();
>                 Console.WriteLine (c);
>             }
>         }
>
>         excelReader.Close();
>     }
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/ExcelDataReader/ExcelDataReader/issues/51#issuecomment-71634215>
> .
>



-- 

Ian Cox
C.T.O.

Release  Mobile

m:            Matrix Studios, 91 Peterborough Road, London, SW6 3BU
w:            http://www.releasemobile.com

Registered in England.  Number 06323661

**Comment by cilerler** [comment]
Thank you for the prompt reply.  Unfortunately, I'm already using the **develop** branch.  :(

**Comment by ExcelDataReader** [comment]
Ah ok. And you have same problem on master branch? i.e. is it an issue that
is in both branches?

On Tue, Jan 27, 2015 at 1:37 PM, Cengiz Ilerler <notifications@github.com>
wrote:

> Thank you for the prompt reply. Unfortunately, I'm already using the
> *develop* branch. :(
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/ExcelDataReader/ExcelDataReader/issues/51#issuecomment-71648252>
> .
>



-- 

Ian Cox
C.T.O.

Release  Mobile

m:            Matrix Studios, 91 Peterborough Road, London, SW6 3BU
w:            http://www.releasemobile.com

Registered in England.  Number 06323661

**Comment by cilerler** [comment]
Nope, didn't confirm on Master yet, will do.  In the mean time, beside encoding I believe there is an another root problem there.  As you can see in screenshots value **28 Ekim 1998 Carsamba** becomes **8 Ekim 1998 Carsamba|**

**Comment by ExcelDataReader** [comment]
Hmmm. I would definitely try master then. This may be a new bug in develop.

On Tue, Jan 27, 2015 at 1:52 PM, Cengiz Ilerler <notifications@github.com>
wrote:

> Nope, didn't confirm on Master yet, will do. In the mean time, beside
> encoding I believe there is an another root problem there. As you can see
> in screenshots value *28 Ekim 1998 Carsamba* becomes *8 Ekim 1998
> Carsamba|*
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/ExcelDataReader/ExcelDataReader/issues/51#issuecomment-71650282>
> .
>



-- 

Ian Cox
C.T.O.

Release  Mobile

m:            Matrix Studios, 91 Peterborough Road, London, SW6 3BU
w:            http://www.releasemobile.com

Registered in England.  Number 06323661

**Comment by cilerler** [comment]
Unfortunately same result on master too...

**Comment by Ian1971** [comment]
I'll try and take a look with the file you posted later.

**Comment by cilerler** [comment]
Thank you for your time!

**Comment by Ian1971** [comment]
I checked the sheet you uploaded and it seems that it is a Excel  5 or 95 file (BIFF 5). Although it does take versions older than BIFF8  into account I don't think the library correctly handles anything older then BIFF8 (>=Excel 97). 

This probably explains why you are getting some strange offset and encoding issues.

I'm having a quick look to see how much work it might be to address

**Comment by cilerler** [comment]
Thank you for the update and explanation.  Please let me know if I can do any help to do the implementation.  I really appreciate your time...  Thank you!

**Comment by Ian1971** [comment]
What .Net framework version are you using?

**Comment by cilerler** [comment]
4.5 & 4.6

**Comment by Ian1971** [comment]
I had a quick go, but managed to break most of the tests. So it certainly isn't a quick job. If you are interested the spec is here. http://www.openoffice.org/sc/excelfileformat.pdf
A good example of the differences is 5.63 Label and also 2.5.2 strings.
There have been a few changes between master and develop to make it adhere more to the spec / simplify the code, but I didn't spend any time on BIFFs < 8. On master it does attempt to do the right thing in some places but it isn't consistently treating BIFF < 8 correctly.

**Comment by cilerler** [comment]
Thank you!!!  I will look into that.

**Comment by ZouLufei** [comment]
Chinese Column xls files do not read correctly. It's the same problem?

**Comment by appel1** [comment]
Not sure if it is the same issue as I have since the link to the sample Excel document is broken. 

In my case it fails to read https://wwwrs.resbank.co.za/webindicators/MonthlyReleaseExcel.aspx?DataType=MRDMA because it uses BIFFRECORDTYPE.LABEL for cells with text and XlsBiffLabelCell doesn't seem to take the record header into account. The offset when reading the string from the record was removed here https://github.com/ExcelDataReader/ExcelDataReader/commit/df381f4efad899f6968b11389b3e4c58b31e432f#diff-dd5b258b48b833168a759427bc81edd5L18, not sure why.

**Comment by andersnm** [comment]
Any chance to test with latest prerelease, or reupload the beforementioned "199811021.xls" for further testing?

**Comment by cilerler** [comment]
Sure you may download it from https://goo.gl/B1kCJY

**Comment by andersnm** [comment]
Can confirm this was fixed by #149
